query/common/src/main/java/org/codice/ddf/admin/common/PrioritizedBatchExecutor.java

```diff
@@ -0,0 +1,242 @@
+                            taskBatch.size());
+                }
+                LOGGER.debug("\tFound no valid results in batch {}.", i + 1);
+            }
+
+            LOGGER.debug("Found no valid responses in all {} batches.", tasks.size());
+            return Optional.empty();
+        } finally {
+            cleanUp();
+        }
+    }
+
+    /**
+     * Start task execution and blocks until the highest priority task batch has returned a valid result
+     * according to the task handler, then cleans up remaining tasks. The current instance of the
+     * {@code PrioritizedBatchExecutor} is not useable after calling {@code getFirst(long, TimeUnit)}.
+     *
+     * @param batchWaitTime amount of time to wait for each batch execution
+     * @param timeUnit      {@code TimeUnit} to use for the {@code batchWaitTime}
+     * @return an {@code Optional} containing a task's result, if there was one
+     */
+    public Optional<T> getFirst(long batchWaitTime, TimeUnit timeUnit) {
+        if (batchWaitTime < 0) {
+            throw new IllegalArgumentException("Batch wait time must be 0 or greater.");
+        }
+
+        if (batchWaitTime == 0) {
+            return getFirst();
+        }
+
+        long batchWaitTimeInMillis = TimeUnit.MILLISECONDS.convert(batchWaitTime, timeUnit);
+
+        try {
+            List<CompletionService<T>> prioritizedCompletionServices =
+                    getPrioritizedCompletionServices();
+
+            for (int i = 0; i < tasks.size(); i++) {
+                List<Callable<T>> taskBatch = tasks.get(i);
+                CompletionService<T> completionService = prioritizedCompletionServices.get(i);
+
+                long batchStartTime = new Date().getTime();
+                long batchEndTime = batchStartTime + batchWaitTimeInMillis;
+                long lastPollTime = batchStartTime;
+                LOGGER.debug("Executing batch {} of {}.", i + 1, tasks.size());
+                for (int j = 0; j < taskBatch.size(); j++) {
+                    try {
+                        long pollTime = batchEndTime - lastPollTime;
+                        Future<T> taskFuture;
+                        if (pollTime > 0) {
+                            LOGGER.debug("\tPolling task result queue for {} milliseconds.",
+                                    pollTime);
+                            taskFuture = completionService.poll(pollTime, TimeUnit.MILLISECONDS);
+                            lastPollTime = new Date().getTime();
+                        } else {
+                            LOGGER.debug(
+                                    "\tWaited {} {} for batch results. Skipping remaining tasks in batch {}.",
+                                    batchWaitTime,
+                                    timeUnit.toString(),
+                                    i + 1);
+                            break;
+                        }
+
+                        Optional<T> result = handleTaskResult(taskFuture);
```

Comment:
✏️ You can extract this line through line 190 into a method and reuse that in your untimed method as well.